Prepare to make postfixOps syntax an error (not just a warning) unless the feature is explicitly enabled

som-snytt opened the pull request:
Require postfixOps opt-in instead of just warning.

Fixes scala/scala-dev#462

som-snytt commented:
After fixing gratuitous postfix in the compiler:

```
/var/folders/2_/xb149z895wb5f1y632xp2bw40000gq/T/sbt_19b03fbd/xsbt/Compat.scala:152: error: postfix operator headOption needs to be enabled
by making the implicit value scala.language.postfixOps visible.
This can be achieved by adding the import clause 'import scala.language.postfixOps'
or by setting the compiler option -language:postfixOps.
See the Scaladoc for value scala.language.postfixOps for a discussion
why the feature needs to be explicitly enabled.
          } headOption
            ^
```

som-snytt committed:
Reduce postfix ops

Usages unguarded by language import. In the same
neighborhood, reduce infix as well.

dwijnand commented:
> Maybe @dwijnand knows if it's possible to set the scalac options for this compilation.

Nope, it's not configurable: https://github.com/sbt/zinc/blob/v1.1.5/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/AnalyzingCompiler.scala#L350

> It's ironic that the compatibility layer uses an incompatible feature.

Hardly fair: you're using the legacy 0.13 version of sbt. Not a problem in sbt 1: https://github.com/sbt/zinc/pull/5/commits/ecf121a7ae7ba99cddad69c91117948174dbfc15#diff-569dc93f697b5a143e904e517b998e1dL124.

dwijnand commented:
> not sure if scalac will upgrade to 1.1.x soon

more reason to land https://github.com/scala/scala/pull/6256

> if needing a 0.13.18 for 2.13 would be OK

quite probably not

sjrd commented:
> I'm not sure if scalac will upgrade to 1.1.x soon or if needing a 0.13.18 for 2.13 would be OK.

Even if scalac migrates to sbt 1.1.x, IIUC this change would still break *users* builds using sbt 0.13.x with Scala 2.13.0-M4+.

som-snytt commented:
Yes, but if 0.13.18 is fixed, users can upgrade without much burden.

I'll look into whether a special dispensation is possible for sbt, which wouldn't be the first time.

som-snytt committed:
Reduce postfix ops in tests

som-snytt commented:
/rebuild

som-snytt commented:
/synch

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -742,21 +742,20 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           def hasOption = settings.language contains featureName
-          val OK = hasImport || hasOption
-          if (!OK) {
+          def dispensation = context.unit.source.file.path.endsWith("xsbt/Compat.scala")
```
I'm worried about doing this during type checking. I would be more comfortable with enabling the postfix ops feature from the start when creating a compiler run for this file.

som-snytt replied:
Roger. Or you might say, "Rojair."

adriaanm reviewed (dismissed):
I support this conceptually, but would like to structure the implementation differently.

 1. first commit should just flip the bit on enableRequired
 2. dispensation for sbt should go earlier in the pipeline, so we hit that path check only once
 3. Ignore new warning in code base || Welcome to rebase central!

adriaanm commented:
Golden tickets need to be handed out before checking them more vigorously: the sbt opt-out commit should go before the "more opt-in" commit. Also please split commits into boyscouting and strictly core changes, especially when we're flipping bits. 

Finally, kill your darlings, as William said (Faulkner, not the man in black): "Make postFixOps more opt-in" is a bit too clever.

som-snytt commented:
This is branch order, but github shows date order. Maybe I broke something. The bit is flipped in its own commit, or do you mean you'd like the required test fixes to precede it?
```
commit d91f4a4bfbd3a9b802ecebb9a9b6044c08eac0db (HEAD -> issue/post-postfix, origin/issue/post-postfix)
Author: Som Snytt <som.snytt@gmail.com>
Date:   Fri Jun 1 15:56:20 2018 -0700

    Make postfixOps more opt-in
    
    Require postfixOps feature opt-in instead of just warning.

commit fdcf987c246bb8db2cfc3f2db9a67cfdfe62128e
Author: Som Snytt <som.snytt@gmail.com>
Date:   Fri Jun 1 17:22:38 2018 -0700

    sbt gets a by

commit b3ca049600577a8227609e747868277708a28cd4
Author: Som Snytt <som.snytt@gmail.com>
Date:   Thu May 3 15:36:53 2018 -0700

    Reduce postfix ops in tests

commit b64946344b8b7b7171952a8beee649f84ac3f871
Author: Som Snytt <som.snytt@gmail.com>
Date:   Wed Apr 25 21:13:58 2018 -0700

    Reduce postfix ops
    
    Usages unguarded by language import. In the same
    neighborhood, reduce infix as well.
```

adriaanm commented:
Ah, right! GH using date order is really unfortunate. Scabot just blindly follows it. I assume there's a rebase option to fix this?

For the boy scouting, I was referring to the boolean algebra wizardry in the bit flip commit. It's not a big deal, but in general it's just nicer for reviewing to separate behavior changes from cleanups. Both are appreciated, just better separate.

som-snytt commented:
Note to self: t5675 is what rang a bell for the double-count issue for `-feature` over on https://github.com/scala/scala/pull/6635

dwijnand commented:
> I assume there's a rebase option to fix this?

I think it's possible to fix this with `--committer-date-is-author-date`.

som-snytt committed:
Make checkfeature more expressionistic

som-snytt committed:
Enable postfix for sbt compat file

If compiling exactly one file with a path ending
in `xsbt/Compat.scala`, enable postfix ops.

This avoids an error when sbt auto-compiles that
file, after postfixOps feature is required.

This commit does not attempt a general mechanism
for adding hacks or hooks for special cases.

som-snytt commented:
I can't tell yet what my mistake was.

som-snytt commented:
The last one is that test doesn't use the new compiler:
```
[test] Waiting for lock on /home/jenkins/.sbt/boot/scala-2.10.7/org.scala-sbt/sbt/0.13.17/sbt.components.lock to be available...
[test] /tmp/sbt_148a1a7/xsbt/Compat.scala:152: error: postfix operator headOption needs to be enabled
[test] by making the implicit value scala.language.postfixOps visible.
[test] This can be achieved by adding the import clause 'import scala.language.postfixOps'
[test] or by setting the compiler option -language:postfixOps.
[test] See the Scaladoc for value scala.language.postfixOps for a discussion
[test] why the feature needs to be explicitly enabled.
[test]           } headOption
[test]             ^
[test] one error found
[test] [error] 14 of 14 test tasks failed:
[test] [error] - junit/test
[test] [error]   - library/compile:compileIncremental failed: Error compiling sbt component 'compiler-interface'
```

som-snytt commented:
/rebuild

adriaanm commented:
Ah right! Hrm, we'll have to release a new starr then, before we flip the bit. Other than that, LGTM -- thanks for finessing the commits.

SethTisue commented:
at https://contributors.scala-lang.org/t/lets-drop-postfix-operators/1457/11?u=sethtisue @LPTK expressed skepticism about the desirability of actually removing this in 2.14.

I think the other changes in this PR take care of disarming the footgun. so maybe the actual removal under `-Xsource:2.14` isn't really necessary? I don't have a strong feeling about it, but I think it's worth a last thought before merge.

SethTisue commented:
note also @sjrd's comments at https://contributors.scala-lang.org/t/lets-drop-postfix-operators/1457/2?u=sethtisue, which @LPTK essential echoed

SethTisue commented:
one more thought: I think the error message should be more explicit that this "feature" is actually undesirable, likely to be removed in a future Scala, and is still supported at all _only for backward compatibility_ and for no other reason.

this isn't a "proceed, but with caution" type thing anymore like `higherKinds` and so son... this one is a "no, really, please don't"

som-snytt committed:
Juicier language for language feature object

Use stronger language to dissuade from postfix,
and make the descriptions more uniform.

Distinguish required and convenience flags in docs.

som-snytt commented:
@SethTisue Please take a look at the doc change in the last commit.

som-snytt commented:
@adriaanm Moved the post-starr commit to the linked PR, in case that's the right thing to do.

SethTisue commented:
> Please take a look at the doc change in the last commit 

LGTM except that I'd suggest appending " (not recommended)" here

```diff
+ *   - [[postfixOps          `postfixOps`]]          enables postfix operator notation
```

som-snytt commented:
OK, I'll add `" (not Seth-approved)"` or `" (not recommended by Seth, who knows)"`, whichever you prefer. I guess I'll know your preference by your reply to this comment. Actually, following Lacan, it should be, `" (not recommended by Seth, the subject supposed to know)"`.

SethTisue commented:
the `-Y` is an other

som-snytt commented:
I just read that NYTimes piece about how the `Y` chromosome is not so useless after all. I assume `-Z` is definitely the big-Oh Other, a status we can grant `scalaZ`, from the contemporary perspective. We could also consider products of `Scala Bridge` under `BridgeZ` pronounced bridges. ("Big-Oh Other" is, IIRC, an ontological distinction, as in `Sein` vs `sein`, nothing to do with `O(n)` big-oh. It's worth advertising this to the community who may not have read this stuff in college.) (Also, did anyone ever pun on `sein` as `his`? I bet they forgot that one. I was just telling the daughter, who wants to learn French and was like what's with vous vs tu, that using "their" for singular is a similar usage. I also mentioned the royal we. We are not amused. But big-Oh we are very much amused. `Oui are amused.`)

som-snytt committed:
Disrecommend postfix more explicitly

som-snytt commented:
@adriaanm @SethTisue if we can move this along, I'll have time to help with community build issues.

I feel no special animus toward postfix, but hoping for clarity.

It needs restarr for the other PR to put the nail in the coffin.

SethTisue commented:
oh, is that because "we'll have to release a new starr"? if so, then 1) can you update the PR title, and 2) yes, let's merge this.

SethTisue commented:
postfix ops are looking *really* nervous now

som-snytt commented:
Postfix ops are the cockroaches of syntax.

hepin1989 commented:
Cool~
